### PR TITLE
SmallRye Keystore

### LINKED ISSUE #11089: Securing credentials/passwords not possible with Quarkus distribution
### Describe the bug

It seems there is currently no way to secure/hide/obfuscate any of the passwords (i.e. db-password, https-key-store-password, https-trust-store-password, referenced in https://www.keycloak.org/server/all-config).

With Wildfly one could use Elytron encrypted values and "credential-stores".

This is an issue because, as @FX-Liagre points out in the referenced discussions

> [...]And our customers' security guys (and DB admins...) tend to faint when you present them a config text file with clear text DB credentials... :-(

### Version

17.0.0

### Expected behavior

Have some sort of encryption ready for hiding credentials or passwords.

### Actual behavior

No such feature available in Quarkus, but still is in Wildfly.

### How to Reproduce?

Provide any clear-text passwords inside the `keycloak.conf`.
Find someone who does a security scan on the installation -> they will find the clear-text password and might possibly faint over it

### Anything else?

See also Discussion:
https://github.com/keycloak/keycloak/discussions/10595

### PR DESCRIPTION
**Pull request by Pepo48**
* added integration with SmallRye Keystore (keycloak#19281)
* added the related properties mapper
* added tests

Signed-off-by: Peter Zaoral <pzaoral@redhat.com>

Closes #11089

### CODE REVIEW

**Comment by Pepo48** [comment]
Adjustments that still need to be done:
- HelpCommandDistTest stabilization
- Documentation

File: docs/guides/server/configuration.adoc [hunk 1]
```diff
@@ -101,6 +103,21 @@ You can also specify an explicit configuration file location using the `[-cf|--c
...
+=== Format for loading properties from a Java KeyStore
+
+Thanks to Keystore Configuration Source you can directly load properties from a Java KeyStore using the `[--config-keystore]` and `[--config-keystore-password]` options.
+Optionally, you can specify the KeyStore type using the `[--config-keystore-type]` option. By default, the KeyStore type is `PKCS12`.
+
+The secrets in a KeyStore needs to be stored using `PBE` (password-based encryption) key algorithm, where a key is derived from a KeyStore password.
+You can generate such a KeyStore using the following `keytool` command:
+[source, bash]
+----
+keytool -importpass -alias kc.log-level -keystore keystore.p12 -storepass keystorepass -storetype PKCS12 -v
```
Comment (vmuzikar) [inline docs/guides/server/configuration.adoc:115]: Maybe a more real life scenario?
```suggestion
keytool -importpass -alias kc.db-password -keystore keystore.p12 -storepass keystorepass -storetype PKCS12 -v
```

File: quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/QuarkusPropertiesDistTest.java [hunk 2]
```diff
@@ -132,6 +133,25 @@ void testUnknownQuarkusBuildTimePropertyApplied(LaunchResult result) {
                 .body(containsString("jvm_gc_"));
     }
 
+    @Test
+    @Launch({ "start", "--http-enabled=true", "--hostname-strict=false", "--config-keystore=keystore" })
+    @Order(10)
+    void testMissingSmallRyeKeyStoreProperty(LaunchResult result) {
+        CLIResult cliResult = (CLIResult) result;
+        cliResult.assertError("The config property smallrye.config.source.keystore.kc-default.password is required but it could not be found in any config source");
```
Comment (vmuzikar) [inline quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/QuarkusPropertiesDistTest.java:141]: This error message brings me to a question whether we shouldn't have some custom validation with more user friendly message. Also, it seems SmallRye doesn't check whether the keystore file actually exists. Maybe we could handle that too. But both are probably just a follow up. WDYT?

CC @pedroigor

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@Pepo48 Thanks, very nice work!

**Comment by vmuzikar** [comment]
@andymunro Could you please review the docs part?

File: docs/guides/server/configuration.adoc [hunk 3]
```diff
@@ -101,6 +103,21 @@ You can also specify an explicit configuration file location using the `[-cf|--c
 
 <@kc.start rootParameters="--config-file=/path/to/myconfig.conf"/>
 
+=== Format for loading properties from a Java KeyStore
+
+Thanks to Keystore Configuration Source you can directly load properties from a Java KeyStore using the `[--config-keystore]` and `[--config-keystore-password]` options.
+Optionally, you can specify the KeyStore type using the `[--config-keystore-type]` option. By default, the KeyStore type is `PKCS12`.
+
+The secrets in a KeyStore needs to be stored using `PBE` (password-based encryption) key algorithm, where a key is derived from a KeyStore password.
```
Comment (andymunro) [inline docs/guides/server/configuration.adoc:111]:
```suggestion
The secrets in a KeyStore need to be stored using the `PBE` (password-based encryption) key algorithm, where a key is derived from a KeyStore password.
```

**Review by andymunro — CHANGES_REQUESTED** [review]
Thanks. Just one change.

File: docs/guides/server/configuration.adoc [hunk 4]
```diff
@@ -101,6 +103,21 @@ You can also specify an explicit configuration file location using the `[-cf|--c
...
+You can generate such a KeyStore using the following `keytool` command:
+[source, bash]
+----
+keytool -importpass -alias kc.log-level -keystore keystore.p12 -storepass keystorepass -storetype PKCS12 -v
+----
+After executing the command, you will be prompted to *Enter the password to be stored*, which represents a value of the `kc.log-level` property above.
+
+For more information about Keystore Configuration Source refer to the https://smallrye.io/smallrye-config/3.2.1/config-sources/keystore/[official SmallRye documentation].
```
[the remark this answers is not in the document]
Comment (Pepo48) [reply docs/guides/server/configuration.adoc]: done

File: docs/guides/server/configuration.adoc [hunk 5]
```diff
@@ -182,7 +199,9 @@ To run a `build`, enter the following command:
 
 This command shows `build options` that you enter. Keycloak distinguishes between **build options**, that are usable when running the `build` command, and **configuration options**, that are usable when starting up the server.
 
-For a non-optimized startup of Keycloak, this distinction has no effect. However, if you run a build before the startup,  only a subset of options is available to the build command. The restriction is due to the build options getting persisted into an optimized Keycloak image. For example, configuration for credentials such as `db-password` must not get persisted for security reasons.
+For a non-optimized startup of Keycloak, this distinction has no effect. However, if you run a build before the startup,  only a subset of options is available to the build command. The restriction is due to the build options getting persisted into an optimized Keycloak image.
+Therefore, we strongly do not recommend to specify any sensitive data (such as `db-password`) as the build options.
+This applies across all the available configuration sources.
```
[the remark this answers is not in the document]
Comment (Pepo48) [reply docs/guides/server/configuration.adoc]: done

File: docs/guides/server/configuration.adoc [hunk 6]
```diff
@@ -101,6 +103,21 @@ You can also specify an explicit configuration file location using the `[-cf|--c
 
 <@kc.start rootParameters="--config-file=/path/to/myconfig.conf"/>
 
+=== Format for loading properties from a Java KeyStore
```
[the remark this answers is not in the document]
Comment (Pepo48) [reply docs/guides/server/configuration.adoc:107]: done

File: quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/QuarkusPropertiesDistTest.java [hunk 7]
```diff
@@ -165,4 +185,13 @@ public void accept(KeycloakDistribution distribution) {
...
     }
+
+    public static class UpdateSmallRyeKeystoreProps implements Consumer<KeycloakDistribution> {
+
+        @Override
+        public void accept(KeycloakDistribution distribution) {
+            distribution.setProperty("config-keystore", Paths.get("src/test/resources/keystore").toAbsolutePath().normalize().toString());
+            distribution.setProperty("config-keystore-password", "secret");
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Pepo48) [reply quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/QuarkusPropertiesDistTest.java]: done

File: docs/guides/server/configuration.adoc (hunk 3 above)
Comment (Pepo48) [reply docs/guides/server/configuration.adoc]: @andymunro thanks, I corrected the sentence.

**Comment by Pepo48** [comment]
@vmuzikar thanks for the review, I incorporated the suggestions in the latest commit.

File: docs/guides/server/configuration.adoc [hunk 8]
```diff
@@ -101,6 +104,19 @@ You can also specify an explicit configuration file location using the `[-cf|--c
...
```
Comment (vmuzikar) [inline docs/guides/server/configuration.adoc:107]: Could you please also add the keystore in the example to be super explicit to users:
https://github.com/keycloak/keycloak/blob/7fccb6ab797c5af8028d0e22fc0243d970d23b8d/docs/guides/server/configuration.adoc?plain=1#L22

File: quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/QuarkusPropertiesDistTest.java (hunk 2 above)
Comment (vmuzikar) [reply quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/QuarkusPropertiesDistTest.java:141]: Actually, even the config file (`-cf`) behaves the same – no error when file does not exist.

**Review by vmuzikar — COMMENTED** [review]
@Pepo48 Thanks for the update! Added one last comment.

File: docs/guides/server/configuration.adoc (hunk 8 above)
Comment (Pepo48) [reply docs/guides/server/configuration.adoc:110]: thanks, suggestion applied

**Comment by Pepo48** [comment]
@vmuzikar I addressed the last bit and squashed the commits.

Thanks for the review.

**Review by vmuzikar — DISMISSED** [review]
@Pepo48 Thank you, LGTM!

File: docs/guides/server/configuration.adoc [hunk 9]
```diff
@@ -101,6 +107,23 @@ You can also specify an explicit configuration file location using the `[-cf|--c
...
```
Comment (pedroigor) [inline docs/guides/server/configuration.adoc:110]: I think we should make this section to be more explicit about `Setting sensitive options using a Java KeyStore` or something like that.

File: quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/QuarkusPropertiesDistTest.java [hunk 10]
```diff
@@ -132,6 +133,26 @@ void testUnknownQuarkusBuildTimePropertyApplied(LaunchResult result) {
...
+    @Order(10)
+    void testMissingSmallRyeKeyStoreProperty(LaunchResult result) {
+        CLIResult cliResult = (CLIResult) result;
+        cliResult.assertError("The config property smallrye.config.source.keystore.kc-default.password is required but it could not be found in any config source");
+        cliResult.assertNoBuild();
+    }
+
+    @Test
+    @Order(11)
+    void testSmallRyeKeyStoreConfigSource(KeycloakDistribution distribution) {
```
Comment (pedroigor) [inline quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/QuarkusPropertiesDistTest.java:147]: Is it actually covering loading the database credentials from the Keystore?

File: docs/guides/server/configuration.adoc [hunk 11]
```diff
@@ -33,11 +34,14 @@ The following example shows how the `db-url` value is set in three configuration
 
 |Configuration file
 |`db-url=confFileValue`
+
+|Java KeyStore file
+|`kc.db-url=keystoreValue`
```
[the remark this answers is not in the document]
Comment (vmuzikar) [reply docs/guides/server/configuration.adoc:39]: We wanted to align it with an existing example which used `db-url`. And I thing a `db-url` could be also considered a sensitive option maybe? ;)

**Comment by pedroigor** [comment]
@Pepo48 We are finally providing something for setting sensitive options !! A very neat integration with SmallRye.

Just a few comments about documentation and testing.

File: quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/QuarkusPropertiesDistTest.java (hunk 10 above)
Comment (vmuzikar) [reply quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/QuarkusPropertiesDistTest.java:147]: Yes, the log level as it is easiest to test.

**Review by pedroigor — APPROVED** [review]

**Review by vmuzikar — APPROVED** [review]

**Pull request merged**